pkg/canary/knative_controller.go

```diff
@@ -0,0 +1,161 @@
+func (kc *KnativeController) SetStatusFailedChecks(cd *flaggerv1.Canary, val int) error {
+	return setStatusFailedChecks(kc.flaggerClient, cd, val)
+}
+
+// SetStatusWeight updates the canary status weight value
+func (kc *KnativeController) SetStatusWeight(cd *flaggerv1.Canary, val int) error {
+	return setStatusWeight(kc.flaggerClient, cd, val)
+}
+
+// SetStatusIterations updates the canary status iterations value
+func (kc *KnativeController) SetStatusIterations(cd *flaggerv1.Canary, val int) error {
+	return setStatusIterations(kc.flaggerClient, cd, val)
+}
+
+// SetStatusPhase updates the canary status phase
+func (kc *KnativeController) SetStatusPhase(cd *flaggerv1.Canary, phase flaggerv1.CanaryPhase) error {
+	return setStatusPhase(kc.flaggerClient, cd, phase)
+}
+
+func (kc *KnativeController) Initialize(cd *flaggerv1.Canary) (bool, error) {
+	if cd.Status.Phase == "" || cd.Status.Phase == flaggerv1.CanaryPhaseInitializing {
+		service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+		}
+
+		service.Annotations["flagger.app/primary-revision"] = service.Status.LatestCreatedRevisionName
+		percent := int64(100)
+		traffic := make([]v1.TrafficTarget, 1)
+		traffic = append(traffic, v1.TrafficTarget{Percent: &percent, RevisionName: service.Status.LatestCreatedRevisionName})
+		service.Spec.Traffic = traffic
+
+		_, err = kc.knativeClient.ServingV1().Services(cd.Namespace).Update(context.TODO(), service, metav1.UpdateOptions{})
+		if err != nil {
+			return true, fmt.Errorf("service %s.%s update query error %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+		}
+	}
+
+	return true, nil
+}
+
+func (kc *KnativeController) Promote(cd *flaggerv1.Canary) error {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	service.Annotations["flagger.app/primary-revision"] = service.Status.LatestCreatedRevisionName
+	_, err = kc.knativeClient.ServingV1().Services(cd.Namespace).Update(context.TODO(), service, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("service %s.%s update query error %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	return nil
+}
+
+func (kc *KnativeController) HasTargetChanged(cd *flaggerv1.Canary) (bool, error) {
+	service, err := kc.knativeClient.ServingV1().Services(cd.Namespace).Get(context.TODO(), cd.Spec.TargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return true, fmt.Errorf("service %s.%s get query error: %w", cd.Spec.TargetRef.Name, cd.Namespace, err)
+	}
+
+	return hasSpecChanged(cd, service.Status.LatestCreatedRevisionName)
```

Comment:
(not a flagger dev)

What 'applies' the new Service `spec.template` that would trigger the rollout?

Reply:
This was the process that I'd envisioned:

1. User creates a Knative service for their application
2. User creates a Flagger configuration pointing at the Knative service
3. Flagger takes over the traffic management of the Knative service
4.a. It adds the metadata it needs to operate
4.b. Wasn't done in this original pull request but it can also probably set the default traffic routing to zero for all revisions & manually set the latest to 100%
5. When the user wants to update an application they update the Knative service like they would usually
6. Flagger watches for the new revision & then does the traffic management changeover